src/main/scala/org/constellation/p2p/PeerAPI.scala

```diff
@@ -143,43 +145,109 @@ class PeerAPI(override val ipManager: IPManager[IO])(
                   .ifM(getInfo, IO.pure(None))
               } {
                 case None => complete(StatusCodes.NotFound)
-                case _    => complete(StatusCodes.OK, Array.empty[Byte])
+                case _    => complete(StatusCodes.OK, Array()) //need Array() here, Array.empty marshalls as List
               }
+          }
+        } ~
+          path(LongNumber) { height =>
+            entity(as[(String, Array[Array[Byte]])]) {
+              case (address, curCheckpointHashes) =>
+                val curSnap = dao.snapshotService.storedSnapshot.get.unsafeRunSync()
+                if (height == curSnap.height) {
+                  val deSerCheckpointHashes =
+                    curCheckpointHashes.flatMap(chunkDeSerialize[Seq[String]](_, "snapshot/info/curCheckpointHashes"))
+                  logger.debug(
+                    s"snapshot/info/${height} from node address: ${address} - height == curSnap.height: true  - deSerCheckpointHashes.size ${deSerCheckpointHashes.size}"
+                  )
+                  val getInfo = dao.snapshotService.getSnapshotInfo.flatMap { info =>
+                    val checkpointsToGet = info.acceptedCBSinceSnapshotHashes.toList.diff(deSerCheckpointHashes.toList)
+                    checkpointsToGet.traverse {
+                      dao.checkpointService.fullData(_)
+                    }.flatMap { cbs =>
+                      val infoSer = info.copy(acceptedCBSinceSnapshotCache = cbs.flatten).toSnapshotInfoSer()
+                      val res = dao.snapshotService.recentSnapshotInfo
+                        .put(address, infoSer)
+                        .flatTap(_ => IO { logger.warn(s"snapshot/info recentSnapshotInfo updating") })
+                        .map(_ => Array.empty[Byte].some)
+                      val testPut = dao.snapshotService.recentSnapshotInfo.lookup(address)
+                      val check = (res >> testPut).unsafeRunSync()
+                      val didSave = (check.get == infoSer)
+                      logger.debug(s"didSave ${didSave} - for address $address with snapshotHashes ${info
+                        .copy(acceptedCBSinceSnapshotCache = cbs.flatten)
+                        .snapshotHashes} - for node ${dao.id.address}")
+                      res
+                    }
+                  }
+                  APIDirective.handle {
+                    dao.cluster.getNodeState
+                      .map(NodeState.canActAsRedownloadSource)
+                      .ifM(getInfo, IO.pure(None))
+                  } {
+                    case None => complete(StatusCodes.NotFound)
+                    case _    => complete(StatusCodes.OK, Array())
+                  }
+                } else {
+                  val heightPrefix = SnapshotService.snapshotInfoFileHeightPrefix(height)
+                  val snapshotInfoDirectories = dao.snapshotInfoPath.collectChildren(_.isDirectory).toList //need to call toList here
```

Comment:
I can see this few times at least. It should be exported to a service which will manage all snapshot info directories/files.

Reply:
I think so too